include/deal.II/physics/transformations.h

```diff
@@ -0,0 +1,1177 @@
+    * @f]
+    *
+    * @param[in] N The referential normal unit vector $\mathbf{N}$
+    * @param[in] F The deformation gradient tensor $\mathbf{F} \left( \mathbf{X} \right)$
+    * @return      The scaled spatial normal vector $\mathbf{n} \frac{da}{dA}$
+    */
+    template<int dim, typename Number>
+    Tensor<1,dim,Number>
+    nansons_formula (const Tensor<1,dim,Number> &N,
+                     const Tensor<2,dim,Number> &F);
+
+//@}
+  };
+
+}
+
+
+
+#ifndef DOXYGEN
+// ------------------------- inline functions ------------------------
+
+namespace internal
+{
+  namespace Physics
+  {
+    template <int dim, typename Number>
+    inline
+    Tensor<1,dim,Number>
+    transformation_contraction (const Tensor<1,dim,Number> &V,
+                                const Tensor<2,dim,Number> &F)
+    {
+      return contract<1,0>(F, V);
+    }
+
+
+
+    template <int dim, typename Number>
+    inline
+    Tensor<2,dim,Number>
+    transformation_contraction (const Tensor<2,dim,Number> &T,
+                                const Tensor<2,dim,Number> &F)
+    {
+      return contract<1,1>(F,contract<1,0>(F, T));
+    }
+
+
+
+    template <int dim, typename Number>
+    inline
+    dealii::SymmetricTensor<2,dim,Number>
+    transformation_contraction (const dealii::SymmetricTensor<2,dim,Number> &T,
+                                const Tensor<2,dim,Number>                  &F)
+    {
+      Tensor<2,dim,Number> tmp_1;
+      for (unsigned int i=0; i<dim; ++i)
+        for (unsigned int J=0; J<dim; ++J)
+          for (unsigned int I=0; I<dim; ++I)
+            tmp_1[i][J] += F[i][I] * T[I][J];
+
+      dealii::SymmetricTensor<2,dim,Number> out;
+      for (unsigned int i=0; i<dim; ++i)
+        for (unsigned int j=i; j<dim; ++j)
+          for (unsigned int J=0; J<dim; ++J)
+            out[i][j] += F[j][J] * tmp_1[i][J];
+
+      return out;
+    }
+
+
+
+    template <int dim, typename Number>
+    inline
+    Tensor<4,dim,Number>
+    transformation_contraction (const Tensor<4,dim,Number> &H,
+                                const Tensor<2,dim,Number> &F)
+    {
+      // Its significantly quicker (in 3d) to push forward
+      // each index individually
+      return contract<1,3>(F,contract<1,2>(F,contract<1,1>(F,contract<1,0>(F, H))));
+    }
+
+
+
+    template <int dim, typename Number>
+    inline
+    dealii::SymmetricTensor<4,dim,Number>
+    transformation_contraction (const dealii::SymmetricTensor<4,dim,Number> &H,
+                                const Tensor<2,dim,Number>                  &F)
+    {
+      // Its significantly quicker (in 3d) to push forward
+      // each index individually
+
+      Tensor<4,dim,Number> tmp_1;
+      for (unsigned int i=0; i<dim; ++i)
+        for (unsigned int J=0; J<dim; ++J)
+          for (unsigned int K=0; K<dim; ++K)
+            for (unsigned int L=0; L<dim; ++L)
+              for (unsigned int I=0; I<dim; ++I)
+                tmp_1[i][J][K][L] += F[i][I] * H[I][J][K][L];
```

Comment:
Yes, that's correct. I think that the optimal approach is for the first and the last of these loops to be manually constructed because of the various tensor symmetries. I can rewrite the middle two using the `contract<>` function since all quantities are standard tensors though.